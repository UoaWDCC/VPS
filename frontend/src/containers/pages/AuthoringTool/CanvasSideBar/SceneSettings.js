import React, { useContext } from "react";
import TextField from "@material-ui/core/TextField";
import { withStyles } from "@material-ui/core/styles";
import { InputAdornment } from "@material-ui/core";
import SceneContext from "../../../../context/SceneContext";

import styles from "../../../../styling/CanvasSideBar.module.scss";

const CustomTextField = withStyles({
  root: {
    marginTop: "0.5em",
    marginBottom: "1.5em",

    "& label.Mui-focused": {
      color: "#0080a7 ",
    },
    "& .MuiInput-underline:after": {
      borderBottomColor: "0080a7",
    },
  },
})(TextField);

/**
 * This component displays the settings of a scene, such as the scene name
 * @component
 */
export default function SceneSettings() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);

  return (
    <>
      <div className={styles.sceneSettingsContainer}>
        <h1 className={styles.sideBarHeader}>Scene Settings</h1>
        <div className={styles.sideBarBody}>
          {/* input for scene name */}
          <CustomTextField
            label="Scene Name"
            value={currentScene?.name}
            fullWidth
            onChange={(event) => {
              setCurrentScene({
                ...currentScene,
                name: event.target.value,
              });
            }}
          />
          {/* input for scene timer duration */}
          <CustomTextField
            label="Scene Timer Duration"
            type="number"
            value={currentScene?.time || ""}
            fullWidth
            onChange={(event) => {
              // limiting scene timer duration
              const timeInput = event.target.value < 0 ? 0 : event.target.value;

              setCurrentScene({
                ...currentScene,
                time: timeInput,
              });
            }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">seconds</InputAdornment>
              ),
            }}
            InputLabelProps={{
              // label moves up whenever there is input
              shrink: currentScene?.time || currentScene?.time === 0,
            }}
          />
        </div>
      </div>
    </>
  );
}
