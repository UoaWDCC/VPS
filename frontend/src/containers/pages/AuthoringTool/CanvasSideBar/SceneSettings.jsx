import {
  InputLabel,
  InputAdornment,
  FormControl,
  FormControlLabel,
  Select,
  Checkbox,
  Typography,
} from "@material-ui/core";
import { useParams } from "react-router-dom";
import { usePut } from "hooks/crudHooks";
import AuthenticationContext from "context/AuthenticationContext";
import TextField from "@material-ui/core/TextField";
import { withStyles } from "@material-ui/core/styles";
import { useContext, useState, useEffect } from "react";
import ScenarioContext from "context/ScenarioContext";
import SceneContext from "../../../../context/SceneContext";

import styles from "../../../../styling/CanvasSideBar.module.scss";
import CustomInputLabelStyles from "./CustomPropertyInputStyles/CustomInputLabelStyles";
import CustomCheckBoxStyles from "./CustomPropertyInputStyles/CustomCheckBoxStyles";

const CustomInputLabel = CustomInputLabelStyles()(InputLabel);
const CustomCheckBox = CustomCheckBoxStyles()(Checkbox);
const CustomTextField = withStyles({
  root: {
    marginTop: "0.5em",
    marginBottom: "1em",

    "& label.Mui-focused": {
      color: "#0080a7 ",
    },
    "& .MuiInput-underline:after": {
      borderBottomColor: "0080a7",
    },
  },
})(TextField);

/**
 * This component displays the settings of a scene, such as the scene name
 * @component
 */
export default function SceneSettings() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);
  const { roleList } = useContext(ScenarioContext);
  const [selectedRoles, setSelectedRoles] = useState([]);

  const [checked, setChecked] = useState([]);
  const [allChecked, setAllChecked] = useState(false);

  useEffect(() => {
    if (currentScene.roles) {
      setSelectedRoles(currentScene.roles);
    }
    if (roleList.length > 0) {
      const initialCheckedState = roleList.map((role) =>
        (currentScene.roles || []).includes(role)
      );
      setChecked(initialCheckedState);

      // Initialize allChecked state
      const initialAllCheckedState = initialCheckedState.every(
        (check) => check
      );
      setAllChecked(initialAllCheckedState);
    }
  }, [roleList, currentScene.roles]);

  const { scenarioId } = useParams();
  const { scenes } = useContext(SceneContext);
  const { getUserIdToken } = useContext(AuthenticationContext);
  async function saveRoles(newRoles) {
    const updatedScenes = scenes.map(({ _id, name, roles: oldRoles }) => {
      if (_id === currentScene._id) {
        const roles = newRoles || oldRoles;
        return {
          _id,
          name,
          roles,
        };
      }
      return { _id, name, roles: oldRoles };
    });
    await usePut(
      `/api/scenario/${scenarioId}/scene/roles`,
      updatedScenes,
      getUserIdToken
    );
  }

  const handleCheckboxChange = (index) => {
    const newChecked = [...checked];
    newChecked[index] = !checked[index];
    setChecked(newChecked);

    const isAllChecked = newChecked.every((isChecked) => isChecked);
    setAllChecked(isAllChecked);

    const updatedSelectedRoles = newChecked.reduce((acc, isChecked, idx) => {
      if (isChecked) {
        acc.push(roleList[idx]);
      }
      return acc;
    }, []);
    setSelectedRoles(updatedSelectedRoles);
    saveRoles(updatedSelectedRoles);
  };

  const handleAllToggle = () => {
    const newAllChecked = !allChecked;
    setAllChecked(newAllChecked);

    const newChecked = newAllChecked
      ? new Array(roleList.length).fill(true)
      : new Array(roleList.length).fill(false);
    setChecked(newChecked);

    const newSelectedRoles = newAllChecked ? roleList : [];
    setSelectedRoles(newSelectedRoles);
    saveRoles(newSelectedRoles);
  };

  return (
    <>
      <div className={styles.sceneSettingsContainer}>
        <h1 className={styles.sideBarHeader}>Scene Settings</h1>
        <div className={styles.sideBarBody}>
          {/* input for scene name */}
          <CustomTextField
            label="Scene Name"
            value={currentScene?.name}
            fullWidth
            onChange={(event) => {
              setCurrentScene({
                ...currentScene,
                name: event.target.value,
              });
            }}
          />
          {/* input for scene timer duration */}
          <CustomTextField
            label="Scene Timer Duration"
            type="number"
            value={currentScene?.time ?? 0}
            fullWidth
            onChange={(event) => {
              // limiting scene timer duration
              const timeInput = event.target.value < 0 ? 0 : event.target.value;

              setCurrentScene({
                ...currentScene,
                time: timeInput,
              });
            }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">seconds</InputAdornment>
              ),
            }}
            InputLabelProps={{
              shrink: true,
            }}
          />
          <FormControl fullWidth className={styles.formControl}>
            <CustomInputLabel>Scene Role(s)</CustomInputLabel>
            <Select
              className={styles.selectInput}
              MenuProps={{
                getContentAnchorEl: null,
              }}
              multiple
              value={selectedRoles}
              onChange={(event) => setSelectedRoles(event.target.value)}
              renderValue={(selected) =>
                selected.length === roleList.length
                  ? "All"
                  : selected.join(", ")
              }
            >
              {roleList && roleList.length > 0 ? (
                <div style={{ display: "flex", flexDirection: "column" }}>
                  <FormControlLabel
                    control={
                      <CustomCheckBox
                        checked={allChecked}
                        onChange={handleAllToggle}
                      />
                    }
                    label="All"
                  />
                  {roleList.map((role, index) => (
                    <FormControlLabel
                      control={
                        <CustomCheckBox
                          checked={checked[index]}
                          onChange={() => handleCheckboxChange(index)}
                        />
                      }
                      label={role}
                    />
                  ))}
                </div>
              ) : (
                <Typography className={styles.menuItem}>
                  This scenario currently does not accommodate roles. Please
                  upload a CSV file first under Manage Groups.
                </Typography>
              )}
            </Select>
          </FormControl>
        </div>
      </div>
    </>
  );
}
