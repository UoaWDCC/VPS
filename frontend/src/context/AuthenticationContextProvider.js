import React, { useState } from "react";
import { signInWithRedirect } from "firebase/auth";
import { useAuthState } from "react-firebase-hooks/auth";
import AuthenticationContext from "./AuthenticationContext";
import { auth, googleProvider } from "../firebase/firebase";
import { useGetSimplified } from "../hooks/crudHooks";

/**
 * This is a Context Provider made with the React Context API
 * AuthenticationContext grants access to functions and variables related to Firebase login
 */
export default function AuthenticationContextProvider({ children }) {
  const [user, loading, error] = useAuthState(auth);

  /**
   * No idToken is stored in state to ensure the non-expired idToken is always used
   * @returns idToken or null if user is not signed in
   */
  async function getUserIdToken() {
    if (user) {
      const token = await user.getIdToken();
      return token;
    }

    return null;
  }

  function signInUsingGoogle() {
    signInWithRedirect(auth, googleProvider);
  }

  function signOut() {
    auth.signOut();
  }

  /**
   * This function determines the role of the current logged in user.
   * @returns the role of user
   */
  function getRole() {
    const [staffList, setStaffList] = useState();
    useGetSimplified("/api/staff", setStaffList);

    let onStaffList = false;
    if (user && getUserIdToken() != null) {
      if (staffList !== undefined) {
        for (let i = 0; i < staffList.length; i += 1) {
          if (staffList[i].firebaseID === user.uid) {
            onStaffList = true;
          }
        }
      }

      if (onStaffList) {
        return "admin";
      }
      return "user";
    }

    return "nonexistent";
  }

  // creating user object with role property
  const VpsUser = {
    firebaseUserObj: user,
    role: getRole(),
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getUserIdToken,
        loading,
        user,
        error,
        signOut,
        signInUsingGoogle,
        VpsUser,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
