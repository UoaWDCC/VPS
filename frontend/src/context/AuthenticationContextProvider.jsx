import { signInWithRedirect, signInWithPopup } from "firebase/auth";
import { useState } from "react";
import { useAuthState } from "react-firebase-hooks/auth";
import { auth, googleProvider } from "../firebase/firebase";
import { useGetSimplified } from "../hooks/crudHooks";
import AuthenticationContext from "./AuthenticationContext";

/**
 * This is a Context Provider made with the React Context API
 * AuthenticationContext grants access to functions and variables related to Firebase login
 */
export default function AuthenticationContextProvider({ children }) {
  const [user, loading, error] = useAuthState(auth);

  /**
   * No idToken is stored in state to ensure the non-expired idToken is always used
   * @returns idToken or null if user is not signed in
   */
  async function getUserIdToken() {
    if (user) {
      const token = await user.getIdToken();
      return token;
    }

    return null;
  }

  function signInUsingGoogle() {
    signInWithPopup(auth, googleProvider);
    // signInWithRedirect(auth, googleProvider);
  }

  function signOut() {
    auth.signOut();
  }

  // getting role from backend
  const [userRole, setUserRole] = useState();
  const userID = user == null ? "null" : user.uid; // this is to avoid null pointer exceptions while confining to hook rules
  useGetSimplified(`/api/staff/${userID}`, setUserRole);

  // creating user object with role property
  const VpsUser = {
    firebaseUserObj: user,
    role: userRole,
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getUserIdToken,
        loading,
        user,
        error,
        signOut,
        signInUsingGoogle,
        VpsUser,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
