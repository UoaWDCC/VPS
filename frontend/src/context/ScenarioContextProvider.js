import React, { useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";

export default function ScenarioContextProvider({ children }) {
  const [currentScenario, setCurrentScenario] = useLocalStorage(
    "currentScenario",
    null
  );
  const [scenarios, setScenarios] = useState();
  const { reFetch } = useGet(`api/scenario`, setScenarios);

  return (
    <ScenarioContext.Provider
      value={{
        scenarios,
        setScenarios,
        reFetch,
        currentScenario,
        setCurrentScenario,
      }}
    >
      {children}
    </ScenarioContext.Provider>
  );
}
