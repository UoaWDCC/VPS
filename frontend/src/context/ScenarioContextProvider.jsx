import { useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";

/**
 * This is a Context Provider made with the React Context API
 * ScenarioContextProvider allows access to scenario info and the refetch function
 */
export default function ScenarioContextProvider({ children }) {
  const [currentScenario, setCurrentScenario] = useLocalStorage(
    "currentScenario",
    null
  );
  const [scenarios, setScenarios] = useState();
  const [assignedScenarios, setAssignedScenarios] = useState();
  const [roleList, setRoleList] = useState();

  const { reFetch } = useGet(`api/scenario`, setScenarios);
  const { reFetch: reFetch2 } = useGet(
    `api/scenario/assigned`,
    setAssignedScenarios
  );

  const { reFetch: reFetch3 } = useGet(
    currentScenario ? `api/group/${currentScenario._id}/roleList` : null,
    currentScenario ? setRoleList : () => {}
  );

  return (
    <ScenarioContext.Provider
      value={{
        scenarios,
        setScenarios,
        reFetch,
        assignedScenarios,
        setAssignedScenarios,
        reFetch2,
        reFetch3,
        currentScenario,
        setCurrentScenario,
        roleList,
      }}
    >
      {children}
    </ScenarioContext.Provider>
  );
}
