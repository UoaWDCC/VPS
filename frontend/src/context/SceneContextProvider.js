import React, { useContext, useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";
import SceneContext from "./SceneContext";

export default function SceneContextProvider({ children }) {
  const { currentScenario } = useContext(ScenarioContext);
  const [scenes, setScenes] = useState([]);
  const [currentScene, setCurrentScene] = useLocalStorage("currentScene", null);
  let getScenes = null;
  if (currentScenario) {
    getScenes = useGet(`api/scenario/${currentScenario._id}/scene`, setScenes);
  }

  return (
    <SceneContext.Provider
      value={{
        scenes,
        setScenes,
        reFetch: getScenes?.reFetch,
        currentScene,
        setCurrentScene,
      }}
    >
      {children}
    </SceneContext.Provider>
  );
}
