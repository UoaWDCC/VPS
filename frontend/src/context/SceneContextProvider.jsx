import { useContext, useEffect, useRef, useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import AuthenticationContext from "./AuthenticationContext";
import ScenarioContext from "./ScenarioContext";
import SceneContext from "./SceneContext";
import { useParams } from "react-router-dom";
import { api } from "../util/api";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import LoadingPage from "../features/status/LoadingPage";
import GenericErrorPage from "../features/status/GenericErrorPage";
import toast from "react-hot-toast";
import useVisualScene from "../features/authoring/stores/visual";
import { getScene } from "../features/authoring/scene/scene";

async function getAllScenes(user, id) {
  const res = await api.get(user, `api/scenario/${id}/scene/all`)
  return res.data.filter(Boolean);
}

function updateScenes(user, id, sceneIds) {
  api.put(
    user,
    `/api/scenario/${id}/scene/reorder`,
    { sceneIds });
}

function deleteScene(user, scenarioId, sceneId) {
  api.delete(user, `/api/scenario/${scenarioId}/scene/${sceneId}`);
};

async function saveScene(user, scenarioId, scene) {
  const components = Object.values(scene.components);
  await api.put(user, `/api/scenario/${scenarioId}/scene/${scene._id}`, { ...scene, components });
}

/**
 * This is a Context Provider made with the React Context API
 * SceneContextProvider allows access to scene info and the refetch function
 */
export default function SceneContextProvider({ children }) {
  const { setCurrentScenario } = useContext(ScenarioContext);
  const { user } = useContext(AuthenticationContext);
  const { scenarioId } = useParams();

  const queryClient = useQueryClient();

  // TODO: this is disgusting
  const scenarioQuery = useQuery({
    queryKey: ["scenario", scenarioId],
    queryFn: () => api.get(user, `api/scenario/${scenarioId}`).then(r => r.data),
    enabled: !!scenarioId,
    onSuccess: (data) => setCurrentScenario(data),
  });

  const scenesQuery = useQuery({
    queryKey: ["scenes", scenarioId],
    queryFn: () => getAllScenes(user, scenarioId),
    enabled: !!scenarioId
  });

  const reorderMutation = useMutation({
    mutationFn: (ids) => updateScenes(user, scenarioId, ids),
    onMutate: async (ids) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], (prev = []) => {
        return ids.map(id => prev.find(s => s._id === id));
      });
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      // queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  const deleteMutation = useMutation({
    mutationFn: (id) => deleteScene(user, scenarioId, id),
    onMutate: async (id) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], prev => prev ? prev.filter(s => s._id !== id) : []);
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  const saveSceneMutation = useMutation({
    mutationFn: (scene) => saveScene(user, scenarioId, scene),
    onMutate: async (scene) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], (prev = []) => {
        const index = prev.findIndex(s => s._id === scene._id);
        return index === -1 ? prev : prev.toSpliced(index, 1, scene);
      })
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  if (scenesQuery.isLoading) {
    return <LoadingPage text="Getting scenes..." />;
  }

  if (scenesQuery.isError) {
    console.error(scenesQuery.error);
    return <GenericErrorPage />;
  }

  return (
    <SceneContext.Provider
      value={{
        scenes: scenesQuery.data,
        reorderScenes: reorderMutation.mutate,
        saveScene: saveSceneMutation.mutate,
        deleteScene: deleteMutation.mutate,
        reFetch: scenesQuery.refetch,
      }}
    >
      {children}
    </SceneContext.Provider>
  );
}
