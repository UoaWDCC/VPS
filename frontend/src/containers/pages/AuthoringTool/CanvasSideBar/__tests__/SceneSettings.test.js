import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import AuthenticationContextProvider from "../../../../../context/AuthenticationContextProvider";
import ScenarioContextProvider from "../../../../../context/ScenarioContextProvider";
import SceneContextProvider from "../../../../../context/SceneContextProvider";
import SceneSettings from "../SceneSettings";

test("Scenario Selection page snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <AuthenticationContextProvider>
        <ScenarioContextProvider>
          <SceneContextProvider>
            <SceneSettings />
          </SceneContextProvider>
        </ScenarioContextProvider>
      </AuthenticationContextProvider>
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
