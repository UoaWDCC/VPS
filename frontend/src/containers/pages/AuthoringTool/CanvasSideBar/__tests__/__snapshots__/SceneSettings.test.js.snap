// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario Selection page snapshot test 1`] = `
<div
  className="sceneSettingsContainer"
>
  <h1
    className="sideBarHeader"
  >
    Scene Settings
  </h1>
  <div
    className="sideBarBody"
  >
    <div
      className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-1 MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Scene Name
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-1 MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Scene Timer Duration
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="number"
          value=""
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            seconds
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
